daemon/graphdriver/windows/windows.go

```diff
@@ -129,61 +226,131 @@ func (d *WindowsGraphDriver) Get(id, mountLabel string) (string, error) {
+func (d *Driver) Put(id string) error {
 	logrus.Debugf("WindowsGraphDriver Put() id %s", id)
 
+	rId, err := d.resolveId(id)
+	if err != nil {
+		return err
+	}
+
 	d.Lock()
 	defer d.Unlock()
 
-	if d.active[id] > 1 {
-		d.active[id]--
-	} else if d.active[id] == 1 {
-		if err := hcsshim.DeactivateLayer(d.info, id); err != nil {
+	if d.active[rId] > 1 {
+		d.active[rId]--
+	} else if d.active[rId] == 1 {
+		if err := hcsshim.UnprepareLayer(d.info, rId); err != nil {
 			return err
 		}
-		delete(d.active, id)
+		if err := hcsshim.DeactivateLayer(d.info, rId); err != nil {
+			return err
+		}
+		delete(d.active, rId)
 	}
 
 	return nil
 }
 
-func (d *WindowsGraphDriver) Cleanup() error {
+func (d *Driver) Cleanup() error {
 	return nil
 }
 
 // Diff produces an archive of the changes between the specified
 // layer and its parent layer which may be "".
-func (d *WindowsGraphDriver) Diff(id, parent string) (arch archive.Archive, err error) {
-	return nil, fmt.Errorf("The Windows graphdriver does not support Diff()")
+func (d *Driver) Diff(id, parent string) (arch archive.Archive, err error) {
+	rId, err := d.resolveId(id)
+	if err != nil {
+		return
+	}
+
+	// Getting the layer paths must be done outside of the lock.
+	layerChain, err := d.getLayerChain(rId)
+	if err != nil {
+		return
+	}
+
+	d.Lock()
+
+	// To support export, a layer must be activated but not prepared.
+	if d.info.Flavour == filterDriver {
+		if d.active[rId] == 0 {
+			if err = hcsshim.ActivateLayer(d.info, rId); err != nil {
+				d.Unlock()
+				return
+			}
+			defer func() {
+				if err := hcsshim.DeactivateLayer(d.info, rId); err != nil {
+					logrus.Warnf("Failed to Deactivate %s: %s", rId, err)
+				}
+			}()
+		} else {
+			if err = hcsshim.UnprepareLayer(d.info, rId); err != nil {
+				d.Unlock()
+				return
+			}
+			defer func() {
+				if err := hcsshim.PrepareLayer(d.info, rId, layerChain); err != nil {
+					logrus.Warnf("Failed to re-PrepareLayer %s: %s", rId, err)
+				}
+			}()
+		}
+	}
+
+	d.Unlock()
+
+	return d.exportLayer(rId, layerChain)
 }
 
 // Changes produces a list of changes between the specified layer
 // and its parent layer. If parent is "", then all changes will be ADD changes.
-func (d *WindowsGraphDriver) Changes(id, parent string) ([]archive.Change, error) {
+func (d *Driver) Changes(id, parent string) ([]archive.Change, error) {
 	return nil, fmt.Errorf("The Windows graphdriver does not support Changes()")
 }
 
 // ApplyDiff extracts the changeset from the given diff into the
 // layer with the specified id and parent, returning the size of the
 // new layer in bytes.
-func (d *WindowsGraphDriver) ApplyDiff(id, parent string, diff archive.Reader) (size int64, err error) {
-	start := time.Now().UTC()
-	logrus.Debugf("WindowsGraphDriver ApplyDiff: Start untar layer")
+func (d *Driver) ApplyDiff(id, parent string, diff archive.Reader) (size int64, err error) {
+	rPId, err := d.resolveId(parent)
+	if err != nil {
+		return
```

Comment:
Go will hit a compiler error if any return values are shadowed.  The return value here is correct, because the line above does not create a new err variable; Go is smart enough to recognize that rPId is a new variable but err is pre-existing.  This wouldn't compile otherwise.